File: test/cuBERT_test.cpp
#include "gtest/gtest.h"
#include <random>
#include <iostream>

#include "cuBERT.h"

class cuBertTest : public ::testing::Test {
protected:
    void SetUp() override {
        cuBERT_initialize();
    }

    void TearDown() override {
        cuBERT_finalize();
    }
};

void random_input(int* input_ids, int8_t* input_mask, int8_t* segment_ids, size_t length) {
    std::random_device r;
    std::default_random_engine e(r());

    std::uniform_int_distribution<int> id_dist(0, 21120);
    std::uniform_int_distribution<int> zo_dist(0, 1);
    for (int i = 0; i < length; ++i) {
        input_ids[i] = id_dist(e);
        input_mask[i] = zo_dist(e);
        segment_ids[i] = zo_dist(e);
    }
}

TEST_F(cuBertTest, compute) {
    int max_batch_size = 128;
    int batch_size = 128;
    int seq_length = 32;

    int input_ids[batch_size * seq_length];
    int8_t input_mask[batch_size * seq_length];
    int8_t segment_ids[batch_size * seq_length];
    float logits[batch_size];
    random_input(input_ids, input_mask, segment_ids, batch_size * seq_length);

    void* model = cuBERT_open("bert_frozen_seq32.pb", max_batch_size, seq_length, 12, 12);
    cuBERT_compute(model, batch_size, input_ids, input_mask, segment_ids, logits);
    cuBERT_close(model);

    std::cout << logits[0] << std::endl;
    std::cout << logits[1] << std::endl;
    std::cout << logits[2] << std::endl;
    std::cout << logits[3] << std::endl;
    std::cout << logits[4] << std::endl;
    std::cout << logits[5] << std::endl;
    std::cout << logits[6] << std::endl;
    std::cout << logits[7] << std::endl;
    std::cout << logits[8] << std::endl;
    std::cout << logits[9] << std::endl;
}

TEST_F(cuBertTest, compute_tokenize) {
    int max_batch_size = 128;
    int batch_size = 2;
    int seq_length = 32;
    float output[batch_size];

    const char* text_a[] = {u8"知乎", u8"知乎"};
    const char* text_b[] = {u8"在家刷知乎", u8"知乎发现更大的世界"};

    void* model = cuBERT_open("bert_frozen_seq32.pb", max_batch_size, seq_length, 12, 12);
    void* tokenizer = cuBERT_open_tokenizer("vocab.txt");

    cuBERT_tokenize_compute(model, tokenizer, batch_size,
                            text_a,
                            text_b,
                            output, cuBERT_LOGITS);

    cuBERT_close_tokenizer(tokenizer);
    cuBERT_close(model);

    std::cout << output[0] << std::endl;
    std::cout << output[1] << std::endl;
}

#ifdef HAVE_CUDA
TEST_F(cuBertTest, compute_tokenize_mf) {
    int max_batch_size = 128;
    int batch_size = 2;
    int seq_length = 32;
    float logits[batch_size];
    cuBERT_Output output;
    output.logits = logits;

    const char* text_a[] = {u8"知乎", u8"知乎"};
    const char* text_b[] = {u8"在家刷知乎", u8"知乎发现更大的世界"};

    void* model = cuBERT_open("bert_frozen_seq32.pb", max_batch_size, seq_length, 12, 12, cuBERT_COMPUTE_HALF);
    void* tokenizer = cuBERT_open_tokenizer("vocab.txt");

    cuBERT_tokenize_compute_m(model, tokenizer, batch_size,
                              text_a,
                              text_b,
                              &output, cuBERT_COMPUTE_HALF, 1);

    cuBERT_close_tokenizer(tokenizer);
    cuBERT_close(model);

    std::cout << logits[0] << std::endl;
    std::cout << logits[1] << std::endl;
}
#endif

File: python/cuBERT_test.py
# -*- coding: UTF-8 -*-
import numpy as np
import libcubert as cuBERT

max_batch_size = 128
batch_size = 2
seq_length = 32

input_ids = np.array([
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31],
    [32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]
], dtype=np.int32)

input_mask = np.array([
    [1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0]
], dtype=np.int8)

segment_ids = np.array([
    [1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0],
    [0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0]
], dtype=np.int8)

output_type=cuBERT.OutputType.LOGITS
compute_type=cuBERT.ComputeType.FLOAT
output = np.zeros([batch_size], dtype=np.float32, order='C')

model = cuBERT.Model("../build/bert_frozen_seq32.pb", max_batch_size, seq_length, 12, 12, 
                     compute_type=compute_type)
model.compute(input_ids, input_mask, segment_ids, output, output_type=output_type)
np.testing.assert_almost_equal([-2.9427543, -1.4876306], output, 5)


text_a = [u"知乎", u"知乎"]
text_b = [u"在家刷知乎", u"知乎发现更大的世界"]
model = cuBERT.Model("../build/bert_frozen_seq32.pb", max_batch_size, seq_length, 12, 12, 
                     compute_type=compute_type,
                     vocab_file="../build/vocab.txt")
model.tokenize_compute(text_a, text_b, output, output_type=output_type)
np.testing.assert_almost_equal([-2.51366, -1.47348], output, 5)
